scipy/sparse/sparsetools/csr.h

```diff
@@ -1232,6 +1232,176 @@ void get_csr_submatrix(const I n_row,
+                   const I Ap[],
+                   const I Aj[],
+                   const T Ax[],
+                   I Bj[],
+                   T Bx[])
+{
+    for(I i = 0; i < n_row_idx; i++){
+        const I row = rows[i];
+        const I row_start = Ap[row];
+        const I row_end   = Ap[row+1];
+        Bj = std::copy(Aj + row_start, Aj + row_end, Bj);
+        Bx = std::copy(Ax + row_start, Ax + row_end, Bx);
+    }
+}
+
+
+/*
+ * Slice rows given as a (start, stop, step) tuple.
+ *
+ * Input Arguments:
+ *   I  start
+ *   I  stop
+ *   I  step
+ *   I  Ap[N+1]    - row pointer
+ *   I  Aj[nnz(A)] - column indices
+ *   T  Ax[nnz(A)] - data
+ *
+ * Output Arguments:
+ *   I  Bj - new column indices
+ *   T  Bx - new data
+ *
+ */
+template<class I, class T>
+void csr_row_slice(const I start,
+                   const I stop,
+                   const I step,
+                   const I Ap[],
+                   const I Aj[],
+                   const T Ax[],
+                   I Bj[],
+                   T Bx[])
+{
+    if (step > 0) {
+        for(I row = start; row < stop; row += step){
+            const I row_start = Ap[row];
+            const I row_end   = Ap[row+1];
+            Bj = std::copy(Aj + row_start, Aj + row_end, Bj);
+            Bx = std::copy(Ax + row_start, Ax + row_end, Bx);
+        }
+    } else {
+        for(I row = start; row > stop; row += step){
+            const I row_start = Ap[row];
+            const I row_end   = Ap[row+1];
+            Bj = std::copy(Aj + row_start, Aj + row_end, Bj);
+            Bx = std::copy(Ax + row_start, Ax + row_end, Bx);
+        }
+    }
+}
+
+
+/*
+ * Slice columns given as an array of indices, pass 1.
+ *
+ * Input Arguments:
+ *   I  n_idx           - number of indices to slice
+ *   I  col_idxs[n_idx] - indices to slice
+ *   I  n_row           - major axis dimension
+ *   I  n_col           - minor axis dimension
+ *   I  Ap[n_row+1]     - indptr
+ *   I  Aj[nnz(A)]      - indices
+ *
+ * Output Arguments:
+ *   I  col_offsets[n_col] - cumsum of index repeats
+ *   I  Bp[n_row+1]        - new indptr
+ *
+ */
+template<class I>
+void csr_column_index1(const I n_idx,
+                       const I col_idxs[],
+                       const I n_row,
+                       const I n_col,
+                       const I Ap[],
+                       const I Aj[],
+                       I col_offsets[],
+                       I Bp[])
+{
+    // bincount(col_idxs)
+    for(I jj = 0; jj < n_idx; jj++){
+        const I j = col_idxs[jj];
+        col_offsets[j]++;
+    }
+
+    // Compute new indptr
+    I new_nnz = 0;
+    Bp[0] = new_nnz;
+    for(I i = 0; i < n_row; i++){
+        for(I jj = Ap[i]; jj < Ap[i+1]; jj++){
+            new_nnz += col_offsets[Aj[jj]];
+        }
+        Bp[i+1] = new_nnz;
+    }
+
+    // cumsum in-place
+    I cumsum = col_offsets[0];
+    for(I j = 1; j < n_col; j++){
+        cumsum += col_offsets[j];
```

Comment:
Is there a reason why you aren't simply doing `col_offsets[j] += col_offsets[j - 1]` inside the loop?

Reply:
Nope, changed.